### PR TITLE
Forward declaration was done starting from hazelcast_client.h

### PR DESCRIPTION
**Pull request by cngzhnp**
There are a lot of nested header files which most of might not be needed to include. So, I clean up a little bit starting from hazelcast_client.h file. If acceptable, I could try a little bit more because it takes too much time build & link it.

### CODE REVIEW

**Comment by OzanCansel** [comment]
Root cause of long compile time is boost library itself. See [here](https://github.com/hazelcast/hazelcast-cpp-client/issues/1034). It is possible to reduce compile time with forward declaration for the files which use boost but are not exposed in header files. But `boost::future<T>` is already exposed in public api and it includes 150-160k lines of code by itself.

So probably forward declaration will not be worthwhile.

However, I will run PR builders and will check your PR. Even it doesn't noticably decrease compile time, it is a good practise to use forward declaration technique wherever it is possible.

**Comment by OzanCansel** [comment]
verify

**Review by OzanCansel — CHANGES_REQUESTED** [review]
It compiles without the lines I suggested to remove. Could you make changes?

File: hazelcast/src/hazelcast/client/spi.cpp [hunk 1]
```diff
@@ -51,6 +52,7 @@
 #include "hazelcast/util/AddressHelper.h"
 #include "hazelcast/util/HashUtil.h"
 #include "hazelcast/util/concurrent/BackoffIdleStrategy.h"
+#include "hazelcast/client/internal/nearcache/NearCacheManager.h"
```
[the remark this answers is not in the document]
Comment (cngzhnp) [reply hazelcast/src/hazelcast/client/spi.cpp:55]: Done.

File: hazelcast/src/hazelcast/client/proxy.cpp [hunk 2]
```diff
@@ -31,6 +31,11 @@
 #include "hazelcast/client/internal/partition/strategy/StringPartitioningStrategy.h"
 #include "hazelcast/util/Util.h"
 #include "hazelcast/client/topic/reliable_listener.h"
+#include "hazelcast/client/spi/impl/ClientPartitionServiceImpl.h"
```
[the remark this answers is not in the document]
Comment (cngzhnp) [reply hazelcast/src/hazelcast/client/proxy.cpp:34]: Done.

File: hazelcast/include/hazelcast/client/ringbuffer/read_result_set.h [hunk 3]
```diff
@@ -17,6 +17,11 @@
 
 #include <vector>
 
+#include <boost/optional.hpp>
+
+#include "hazelcast/client/serialization/serialization.h"
```
[the remark this answers is not in the document]
Comment (cngzhnp) [reply hazelcast/include/hazelcast/client/ringbuffer/read_result_set.h:22]: Done.

File: hazelcast/include/hazelcast/client/transaction_context.h [hunk 4]
```diff
@@ -23,6 +28,7 @@
 #include "hazelcast/client/transactional_multi_map.h"
 #include "hazelcast/client/transactional_list.h"
 #include "hazelcast/client/transactional_set.h"
+#include "hazelcast/client/spi/ClientContext.h"
```
[the remark this answers is not in the document]
Comment (cngzhnp) [reply hazelcast/include/hazelcast/client/transaction_context.h:31]: Done.

File: hazelcast/include/hazelcast/client/reliable_topic.h [hunk 5]
```diff
@@ -32,6 +32,9 @@
 #endif
 
 namespace hazelcast {
+namespace util {
+class hz_thread_pool;
+}
```
[the remark this answers is not in the document]
Comment (cngzhnp) [reply hazelcast/include/hazelcast/client/reliable_topic.h:37]: Done.

File: hazelcast/include/hazelcast/client/proxy/ReplicatedMapImpl.h [hunk 6]
```diff
@@ -21,6 +21,9 @@
 #include "hazelcast/client/client_config.h"
 #include "hazelcast/client/serialization/pimpl/data.h"
 #include "hazelcast/client/internal/nearcache/NearCache.h"
+#include "hazelcast/client/internal/nearcache/NearCacheManager.h"
+#include "hazelcast/client/protocol/codec/codecs.h"
+#include "hazelcast/client/entry_event.h"
```
[the remark this answers is not in the document]
Comment (cngzhnp) [reply hazelcast/include/hazelcast/client/proxy/ReplicatedMapImpl.h:26]: Done.

File: hazelcast/include/hazelcast/client/impl/hazelcast_client_instance_impl.h [hunk 7]
```diff
@@ -63,37 +38,51 @@
 #endif
 
 namespace hazelcast {
+class logger;
 namespace client {
 class hazelcast_client;
+class imap;
+class reliable_topic;
```
[the remark this answers is not in the document]
Comment (cngzhnp) [reply hazelcast/include/hazelcast/client/impl/hazelcast_client_instance_impl.h:45]: Done.

**Comment by cngzhnp** [comment]
@OzanCansel Could you please verify that your changes are completed?

**Comment by OzanCansel** [comment]
Thanks for changes, I will run PR builders again.

verify

**Comment by OzanCansel** [comment]
verify

**Review by akeles85 — COMMENTED** [review]
Changes seem ok.

I want to share my results with compile time of library without tests and examples. In my laptop, the compile time takes like below:

1 job:
    Before change: 1 minute 25 seconds
    After change: 1 minute 24 seconds

6 paralel jobs (-j 6):
    Before change: 20 seconds
    After change: 20 seconds

By the way, precompiled headers might be also a good option to consider.

**Comment by OzanCansel** [comment]
verify

**Review by OzanCansel — APPROVED** [review]

**Pull request merged**